packages/internal-testing-tools/android/src/main/kotlin/com/datadog/reactnative/internaltesting/DdInternalTestingImplementation.kt

```diff
@@ -0,0 +1,240 @@
+    }
+
+    override fun updateFeatureContext(
+        featureName: String,
+        updateCallback: (context: MutableMap<String, Any?>) -> Unit
+    ) {
+        core.updateFeatureContext(featureName, updateCallback)
+    }
+
+    override fun getFeatureContext(featureName: String): Map<String, Any?> {
+        return core.getFeatureContext(featureName)
+    }
+
+    // endregion
+
+    // region SdkCore
+
+    override val service: String
+        get() {
+            return core.service
+        }
+
+    override val time: TimeInfo
+        get() {
+            val nanos = System.nanoTime()
+            return TimeInfo(
+                deviceTimeNs = nanos,
+                serverTimeNs = nanos,
+                serverTimeOffsetMs = 0L,
+                serverTimeOffsetNs = 0L
+            )
+        }
+
+    override fun setUserInfo(
+        id: String?,
+        name: String?,
+        email: String?,
+        extraInfo: Map<String, Any?>
+    ) {
+        core.setUserInfo(id, name, email, extraInfo)
+    }
+
+    // endregion
+}
+
+class StubFirstPartyHostHeaderTypeResolver :
+    FirstPartyHostHeaderTypeResolver {
+
+    // region FirstPartyHostHeaderTypeResolver
+
+    override fun isEmpty(): Boolean = false
+
+    override fun isFirstPartyUrl(url: HttpUrl): Boolean = true
+
+    override fun isFirstPartyUrl(url: String): Boolean = true
+
+    override fun headerTypesForUrl(url: String): Set<TracingHeaderType> = setOf(TracingHeaderType.TRACECONTEXT)
+
+    override fun headerTypesForUrl(url: HttpUrl): Set<TracingHeaderType> = setOf(TracingHeaderType.TRACECONTEXT)
+
+    override fun getAllHeaderTypes(): Set<TracingHeaderType> = setOf(TracingHeaderType.TRACECONTEXT)
+
+    // endregion
+}
+
+
+
+internal class FeatureScopeInterceptor(
+    private val featureScope: FeatureScope,
+    private val core: InternalSdkCore,
+) : FeatureScope by featureScope {
+    private val eventsBatchInterceptor = EventBatchInterceptor()
+
+    fun eventsWritten(): List<String> {
+        return eventsBatchInterceptor.events
+    }
+
+    fun clearData() {
+        eventsBatchInterceptor.clearData()
+    }
+
+    // region FeatureScope
+
+    override fun withWriteContext(
+        forceNewBatch: Boolean,
+        callback: (DatadogContext, EventBatchWriter) -> Unit
+    ) {
+        featureScope.withWriteContext(forceNewBatch, callback)
+
+        core.getDatadogContext()?.let {
+            callback(it, eventsBatchInterceptor)
+        }
+    }
+
+    // endregion
+}
+
+
+internal class EventBatchInterceptor: EventBatchWriter {
+    internal val events = mutableListOf<String>()
+
+    override fun currentMetadata(): ByteArray? {
+        TODO("Not yet implemented")
+    }
+
+    fun clearData() {
+        events.clear()
+    }
+
+    override fun write(
+        event: RawBatchEvent,
+        batchMetadata: ByteArray?
+    ): Boolean {
+        val eventContent = String(event.data)
+
+        events.add(events.size,
+            eventContent
+        )
```

Comment:
write is always to the tail by default

```suggestion
        events += eventContent
```